Add logging support to HTTP client

nezihyigitbasi opened the pull request:
Here is a snippet:
```
2018-02-05T10:52:07.675-08:00   http    POST    http://[2401:db00:3020:b19a:face:0:27:0]:7777/v1/task/20180205_185203_00028_53xzu.3.66?summarize        200     1493    6437
2018-02-05T10:52:07.797-08:00   http    GET     http://[2401:db00:3020:b19a:face:0:27:0]:7777/v1/task/20180205_185200_00013_53xzu.3.72?summarize        200     1492    6316
2018-02-05T10:52:07.891-08:00   http    GET     http://[2401:db00:3020:b19a:face:0:27:0]:7777/v1/task/20180205_185203_00027_53xzu.3.35/status   200     478     6222
2018-02-05T10:52:07.998-08:00   http    POST    http://[2401:db00:3020:b19a:face:0:27:0]:7777/v1/task/20180205_185203_00027_53xzu.3.35?summarize        200     1492    6115
2018-02-05T10:52:08.592-08:00   http    GET     http://[2401:db00:3020:b19a:face:0:27:0]:7777/v1/task/20180205_185202_00020_53xzu.1.92?summarize        200     1584    5521
2018-02-05T10:52:08.607-08:00   http    GET     http://[2401:db00:3020:b19a:face:0:27:0]:7777/v1/task/20180205_185158_00005_53xzu.1.90?summarize        200     1582    5506
2018-02-05T10:52:09.164-08:00   http    POST    http://[2401:db00:3020:b19a:face:0:27:0]:7777/v1/task/20180205_185202_00026_53xzu.1.26?summarize        200     1582    4950
2018-02-05T10:52:09.284-08:00   http    GET     http://[2401:db00:3020:b19a:face:0:27:0]:7777/v1/task/20180205_185201_00017_53xzu.1.24/status   200     478     4830
2018-02-05T10:52:09.288-08:00   http    POST    http://[2401:db00:3020:b19a:face:0:27:0]:7777/v1/task/20180205_185201_00017_53xzu.1.24?summarize        200     1584    4826
2018-02-05T10:52:04.109-08:00   http    GET     http://[2401:db00:3020:b207:face:0:33:0]:7777/v1/task/20180205_185202_00023_53xzu.0.0/status    TIMEOUT -1      10004
2018-02-05T10:52:10.239-08:00   http    GET     http://[2401:db00:3020:b196:face:0:23:0]:7777/v1/task/20180205_185206_00047_53xzu.1.47/status   200     478     3884
2018-02-05T10:52:13.608-08:00   http    GET     http://[2401:db00:3020:51a8:face:0:23:0]:7777/v1/task/20180205_185158_00003_53xzu.1.3?summarize 200     1598    515
2018-02-05T10:52:11.957-08:00   http    GET     http://[2401:db00:3020:5165:face:0:2d:0]:7777/v1/task/20180205_185159_00008_53xzu.0.0?summarize 200     1586    2166
2018-02-05T10:52:10.282-08:00   http    GET     http://[2401:db00:3020:b196:face:0:23:0]:7777/v1/task/20180205_185205_00038_53xzu.1.54/status   200     478     3841
2018-02-05T10:52:13.256-08:00   http    GET     http://[2401:db00:3020:b1a0:face:0:21:0]:7777/v1/task/20180205_185159_00007_53xzu.1.55/status   200     480     867
```

electrum commented:
Can you add a section to `http-client/README.md` that documents the log format?

nezihyigitbasi commented:
Added the log format to `http-client/README.md`

nezihyigitbasi commented on http-client/README.md:
````diff
@@ -39,3 +39,18 @@ Using aliases is similar to using filters:
+```
+A request log line has seven columns delimited with the tab character:
+- Request timestamp in the ISO 8601 format.
+- The protocol (http, https).
````
Maybe we should have the version (http 1.1/2.0) instead of the `protocol` column.

dain commented on http-client/src/main/java/io/airlift/http/client/jetty/HttpClientLogLayout.java:
```diff
@@ -0,0 +1,35 @@
+                .append(event.getResponseSize())
+                .append('\t')
+                .append(event.getTimeToLastByte())
+                .append('\n');
```
should we add trace token?  It is passed as a header, so it should be available when constructing the event.

dain commented on http-client/src/main/java/io/airlift/http/client/jetty/HttpClientLogger.java:
```diff
@@ -0,0 +1,98 @@
+    class ResponseInfo
+    {
+        @Nullable
+        private final Response response;
```
use `Optional`

dain commented on http-client/src/main/java/io/airlift/http/client/jetty/HttpClientLogger.java:
```diff
@@ -0,0 +1,98 @@
+            return new ResponseInfo(response, responseSize, null);
+        }
+
+        public static ResponseInfo failed(@Nullable Response response, @Nonnull Throwable failureCause)
```
use `Optional`

dain commented on http-client/src/main/java/io/airlift/http/client/jetty/HttpRequestEvent.java:
```diff
@@ -0,0 +1,177 @@
+
+    public String getResponseCode()
+    {
+        return failureReason.orElseGet(() -> Integer.toString(responseCode));
```
I find this strange.  It means the response code column in the log can be a non numeric value.

nezihyigitbasi replied:
Yes. For example, for requests that fail with a timeout exception that column will be `TIMEOUT` as we don't have a response code. What do we want to have in the logs for such failed requests? Maybe we can have both response code and failure reason columns, and for cases like these former can be empty, etc.

nezihyigitbasi replied:
Keeping it as is as we discussed offline.

dain commented on http-client/src/main/java/io/airlift/http/client/jetty/HttpRequestEvent.java:
```diff
@@ -0,0 +1,177 @@
+            return Optional.empty();
+        }
+
+        String className = failure.getClass().getSimpleName().toUpperCase();
```
set locale

dain commented on http-client/src/main/java/io/airlift/http/client/HttpClientConfig.java:
```diff
@@ -444,4 +451,65 @@ public HttpClientConfig setTimeoutConcurrency(int timeoutConcurrency)
+    }
+
+    @Config("http-client.log.path")
+    public HttpClientConfig setLogPath(String logPath)
```
Add a config description explaining that this the files are named `<client-name>-http...`

dain commented on http-client/src/test/java/io/airlift/http/client/jetty/TestHttpClientLogger.java:
```diff
@@ -0,0 +1,584 @@
+        file = File.createTempFile(getClass().getName(), ".log");
+    }
+
+    @AfterClass(alwaysRun = true)
```
I believe you want `AfterMethod` to match the `BeforeMethod` above

dain commented on http-client/src/test/java/io/airlift/http/client/jetty/TestHttpClientLogger.java:
```diff
@@ -0,0 +1,584 @@
+    public void tearDown()
+            throws IOException
+    {
+        if (!file.delete()) {
```
Do you want to check `exists`?

nezihyigitbasi commented:
Updated the log format and added the protocol version:

```
2018-02-21T18:26:50.438-08:00	HTTP/1.1	DELETE	http://127.0.0.1:8080/v1/task/20180222_022649_00001_dn54g.0.0/results/0	204	0	16	null
```

nezihyigitbasi commented:
@dain any other comments?

martint commented on http-client/src/main/java/io/airlift/http/client/jetty/HttpRequestEvent.java:
```diff
@@ -0,0 +1,180 @@
+    }
+
+    @VisibleForTesting
+    static Optional<String> getFailureReason(ResponseInfo responseInfo)
```
What are possible failure reasons? Have you looked at whether the common cases produces something reasonable given this logic?

nezihyigitbasi replied:
I remember seeing TIMEOUT, CANCELLATION, and IO in the logs.

nezihyigitbasi replied:
So, these are probably good for debugging many cases. If we find a better way we can improve it in the future.

martint commented on http-client/src/test/java/io/airlift/http/client/jetty/TestHttpClientLogger.java:
```diff
@@ -0,0 +1,559 @@
+/*
+ * Copyright 2010 Proofpoint, Inc.
```
Why Proofpoint? This looks like brand new code.

martint commented on http-client/src/main/java/io/airlift/http/client/jetty/NoopLogger.java:
```diff
@@ -0,0 +1,15 @@
+package io.airlift.http.client.jetty;
```
Missing license header

martint commented on http-client/src/main/java/io/airlift/http/client/jetty/HttpRequestEvent.java:
```diff
@@ -0,0 +1,180 @@
+/*
+ * Copyright 2010 Proofpoint, Inc.
```
Why Proofpoint? This looks like brand new code

martint commented on http-client/src/main/java/io/airlift/http/client/jetty/HttpClientLogger.java:
```diff
@@ -0,0 +1,96 @@
+package io.airlift.http.client.jetty;
```
Missing license headers

martint commented on http-client/src/main/java/io/airlift/http/client/jetty/HttpClientLogLayout.java:
```diff
@@ -0,0 +1,37 @@
+package io.airlift.http.client.jetty;
```
Missing license header

nezihyigitbasi commented:
Updated the license headers.
